### PR TITLE
GH-79 - Add support for highlighting search results in the minimap.

### LINKED ISSUE #79: Option to highlight search matches in minimap
## Check list

- [x] I have read through the [README](https://github.com/wfxr/minimap.vim/blob/master/README.md) (especially F.A.Q section)
- [x] I have searched through the existing issues

## Solution you'd like

The ability to configure highlight regex match locations in minimap. 
For example, if I have some function definition I am searching through the document for with ~/~ , In addition to the highlighted search in the buffer I would like to see the corresponding lines in the minimap be highlighted

## Additional context

Here is an illustration
MinimapExample

### PR DESCRIPTION
**Pull request by ZNielsen**
Search results are scaled and positioned to be roughly where they appear
in the buffer. Added a new function `highlight_span` - essentially just
a wrapper around `matchaddpos()`. Allows for highlighting part of a line
instead of the whole line.

This change includes some refactoring to the git_color work:
- The git coloration was moved to a dedicated function
- ~~`matchid`s are now doled out from a common function~~ `matchid`s are now managed by a function for each group.
    - New system supports an arbitrary number of optional color settings

A couple general changes were included to help scaling up multiple discrete color settings:
- Now have settings for group color priority. 
    - ~~Not planning on exposing those unless someone asks for it -- it was hidden before, it remains a back end setting, just now easier to tweak.~~ Someone immediately asked, so it's documented now.

### Big general changes
I changed all set_color functions to use `matchaddpos()`. The vimscript documentation calls out the fact that it is faster than `matchadd()`, as it doesn't have to match against a regex. Since we are dealing with line numbers instead of patterns, it is faster and more idiomatic to use `matchaddpos()`. If you would prefer this change in its own PR, I can break it out.

Closes #79 

## Check list

- [X] I have read through the [README](https://github.com/wfxr/minimap.vim/blob/master/README.md) (especially F.A.Q section)
- [X] I have searched through the existing issues or pull requests
- [x] I have performed a self-review of my code and commented hard-to-understand areas
- [x] I have made corresponding changes to the documentation (when necessary)
    - [x] README.md
    - [x] Vim documentation

## Type of change

- [ ] Bug fix
- [X] New feature
- [X] Improvement of existing features - Improved the accuracy of git color stuff while I was in the area
- [X] Refactor
- [ ] Breaking change
- [x] Documentation change
- [ ] CI / CD

## Test environment

- OS
    - [X] Linux
    - [X] Mac OS X
    - [ ] Windows
    - [ ] Others:
- Vim
    - [X] Neovim: 0.5.0 nightly
    - [ ] Vim: <Version>

### CODE REVIEW

**Comment by ZNielsen** [comment]
Putting the work up for review. Still need to do 2 TODOs, clean up, and add documentation before it's ready for merge

File: autoload/minimap/vim.vim [hunk 1]
```diff
@@ -367,115 +371,298 @@ function! s:update_highlight() abort
...
+    " Color lines, creating a new id for each group
+    for a_span in mm_spans
+        " span_list item: [line_number, column_number, length]
+        call add(g:minimap_search_id_list, s:get_next_matchid())
+        call s:set_span_color(g:minimap_search_color, [a_span],
+            \ g:minimap_search_color_priority, g:minimap_search_id_list[-1], a:winid)
+    endfor
+endfunction
+
+function! s:minimap_color_git(winid, total_lines, mmheight) abort
```
Comment (ZNielsen) [inline autoload/minimap/vim.vim:542]: Same functionality, just moved to a function

**Comment by ZNielsen** [comment]
Rebased into one commit. Code is ready, just need to do the documentation. Should be able to get to that this weekend.

**Comment by ZNielsen** [comment]
More testing has revealed there is sometimes an issue getting the window number when searching with `*` (and I assume `#`). I will look into that when I get the chance.

**Comment by wfxr** [comment]
@ZNielsen Sorry for the late reply. And thank you very much for doing so much work for this. I found a small problem after testing. This pr seams break the vertical span highlight feature (#62, #63). 

image

The position of the highlight region is not correct wether I enable `g:minimap_highlight_search` or not.

**Comment by ZNielsen** [comment]
TODO: Per #83, add the hidden priority options to the documentation.

**Comment by ZNielsen** [comment]
> The position of the highlight region is not correct wether I enable `g:minimap_highlight_search` or not.

@wfxr This might be a platform issue, the feature works for me in Vim8 on MacOS

**Comment by wfxr** [comment]
@ZNielsen I tested both in vim8.2 and neovim 5.0 on Arch Linux. I guess this might not be related the platform. And this problem only occurs when there are closed folds in buffer:

image

You can download this file and test: https://gist.github.com/wfxr/794760e9a2d066559dae252b0a3a1086#file-minimap-vim-pr81-test-go (open it and close all folds by `zM`, then move cursor to the bottom and you will see the highlight position in minimap is not correct)

**Comment by wfxr** [comment]
Another problem I found is nvim always crashes when I resize the window:

```
nvim: /build/neovim-git/src/neovim-git/src/nvim/message.c:2281: msg_scroll_flush: Assertion `pos_delta >= 0' failed.
```

vim will  crash but feels a bit

**Comment by ZNielsen** [comment]
@wfxr Ah, I did not try with folds. Thanks for the details, I'll look into both those issues.

**Comment by ZNielsen** [comment]
@wfxr can I get some more details on the changing window size issue? It seem to work for me when I had 4 windows  up (including the minimap) and I resized them with the mouse/keyboard.

**Comment by wfxr** [comment]
@ZNielsen Thank you, now it works fine with folds.

> @wfxr can I get some more details on the changing window size issue?

This issue only exists when I disable `minimap_highlight_search`. And some error messages are displayed when vim starts:

```
Error detected while processing function minimap#vim#MinimapOpen[1]..<SNR>138_open_window[78]..<SNR>138_update_highlig
ht[1]..<SNR>138_get_window_info:
line    4:
E121: Undefined variable: g:minimap_getting_window_info
E15: Invalid expression: g:minimap_getting_window_info == 0
Error detected while processing function <SNR>138_handle_autocmd[18]..<SNR>138_buffer_enter_handler[4]..<SNR>138_sourc
e_buffer_enter_handler[2]..<SNR>138_update_highlight[1]..<SNR>138_get_window_info:
line    4:
E121: Undefined variable: g:minimap_getting_window_info
E15: Invalid expression: g:minimap_getting_window_info == 0
```

Maybe you forgot to initialize `g:minimap_getting_window_info` in this scenerio?

**Comment by ZNielsen** [comment]
Fixed that missing definition if search highlighting is disabled.

I still can't recreate the window sizing issue. Is this a window within vim, or the terminal window? Would it be possible to get a gif of the crash?

**Comment by wfxr** [comment]
@ZNielsen https://github.com/wfxr/minimap.vim/commit/e74110c21fc0df2af670c0dd275b1e2d3d90d1d4 fixes the resizing issue. I meant the crash may be related to those error message when I said

> This issue only exists when I disable minimap_highlight_search. And some error messages are displayed when vim starts

Sorry I didn't express it clearly.

> Would it be possible to get a gif of the crash?

Sure! Gif before https://github.com/wfxr/minimap.vim/commit/e74110c21fc0df2af670c0dd275b1e2d3d90d1d4 (resizing -> pressing h/l/j/k or whatever -> crash):

vim-crash

**Comment by ZNielsen** [comment]
@wfxr Thanks! I've really enjoyed working on this project, and I appreciate how welcoming you have been.

That was the last issue that I was aware of, so whenever you are done reviewing I think it will be good to merge.

**Review by wfxr — APPROVED** [review]

**Pull request merged**